File: go/tree/1008. 前序遍历构造二叉搜索树.go
package tree

import "sort"

func bstFromPreorder(preorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}
	inorder := make([]int, len(preorder))
	copy(inorder, preorder)
	sort.Ints(inorder)
	pre_val := preorder[0]
	node := &TreeNode{Val: pre_val}
	index := 0
	for i, v := range inorder {
		if v == pre_val {
			index = i
			break
		}
	}
	node.Left = bstFromPreorder(preorder[1 : index+1])
	node.Right = bstFromPreorder(preorder[index+1:])
	return node
}

File: python/树/1008. 前序遍历构造二叉搜索树.py
from typing import List
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if not preorder:
            return
        inorder = sorted(preorder)
        pre_value = preorder[0]
        node = TreeNode(pre_value)
        index = 0
        for i,v in enumerate(inorder):
            if v == pre_value:
                index = i
                break
        node.left = self.bstFromPreorder(preorder[1:index+1])
        node.right = self.bstFromPreorder(preorder[index+1:])
        return node